skylog/storage/cloudfunctions/worker.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package cloudfunctions wraps Merkle tree build worker into a Cloud Function.
+package cloudfunctions
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+
+	"cloud.google.com/go/spanner"
+	"github.com/google/trillian/merkle/compact"
+	"github.com/google/trillian/merkle/rfc6962"
+	"github.com/google/trillian/skylog/core"
+	cs "github.com/google/trillian/skylog/storage/cloudspanner"
+)
+
+const spannerEnvVar = "SPANNER_DB"
+
+var (
+	hasher  = rfc6962.DefaultHasher
+	factory = &compact.RangeFactory{Hash: hasher.HashChildren}
+
+	client *spanner.Client
+)
+
+// BuildMessage is the payload of a Merkle tree building Cloud Pub/Sub event.
+type BuildMessage struct {
+	// Data contains a JSON-encoded BuildJob.
+	// TODO(pavelkalinnikov): Consider protobuf instead.
+	Data []byte `json:"data"`
+}
+
+// BuildJob describes a Merke tree building job.
+type BuildJob struct {
+	TreeID int64  `json:"tree_id"`
+	Begin  uint64 `json:"begin"`
```

Comment:
Is `End` inclusive or exclusive?
Would `Start`, `Num` be clearer?

Reply:
Clarified this in the type comment. I prefer `[begin, end)` when possible, which I believe to be a commonly recognized convention (I might be biased towards people who heard about C++). Compact range is also designed this way.